### ISSUE TITLE
[feature] Scene replication. 

### ISSUE THREAD

**Issue by blueforesticarus** [projectharmonia/bevy_replicon#376]
My (rough draft) proposal for replication of scene children (and other client spawned entities).

Generic case:
1. Children to be replicated are given a ReplicationId component (on both client and server) containing a deterministic ID, in addition to Replicated
2. ReplicationId is included when replicating server entities to the client
3. replicated entities with a ReplicationId component are buffered by the client until the client-side entity with the same ReplicationId become available, at which point a *client side owned* entity map (as opposed to the one currently owned by the server) is updated to point the server entity to the client entity, and the buffered replication messages are applied.
4. duplicate ReplicationId's can be supported for symmetric cases (ie. spawning 10 pigs)

Scene case:
1. Children to be replicated should have AutoReplicationId component attached by client and server
2. a system should be written to replace AutoReplicationId with a ReplicationId based on the hash of the hierarchy path using Name components, using as the root the nearest parent which is *already replicated.* 
ie. `123/my/scene/ball_12`  where 123 is the server entity_id of a replicated entity ( most likely with a MySceneSpawner or such component )

Potential issues: 
- AutoReplicationId might end up with race condition if spawning replicated children of replicated children (which entity is used as the root)
- what should the server do if ReplicationId is added to an already replicated entity.

Prior art: 
I haven't been able to find anything specific to scene replication.

Inspiration for using a deterministic hash for mapping comes from how lightyear does prespawning.
https://docs.rs/lightyear/latest/lightyear/client/prediction/prespawn/struct.PreSpawnedPlayerObject.html

**Comment by Shatur** [comment]
Adding `ReplicationId` to each entity will significantly increase the replication size. I also not sure how adding it to each entity solves the problem.

### The problem

The issue is that if you attach `Replicated` to a scene entity, it will be replicated as a new entity. But we want to map it to already existing entity.

### Possible solution

1. Detect if the entity is a part of a replicated scene.
2. If server adds `Replicated`, send information for client to map it to already spawned scene entity.

But what is scene wasn't spawned by client yet? :thinking: 

### Mapping could be encoded in the replication message like this

1. Server `Entity` of the root scene instance.
2. Array of `Entity` to `Entity` mappings:
  2.1 Local scene `Entity` (each entity in the scene have it, basically ID before spawning).
  2.2 Server spawned `Entity`.

1 and 2.2 will be mapped into client-local entities and 2.1 already known by the client.

### Prior art

Both Unreal Engine and Godot allow to enable replication for every actor or node in a scene. It would be nice to look into their approach.

**Comment by blueforesticarus** [comment]
I was not suggesting adding ReplicationId to every entity, only descendant entities which you want to be replicated to an entity spawned by the client. 

I think this issue is more general than scenes. It applies to any hierarchical structures.

**Comment by Shatur** [comment]
Ah, got it!

I would consider hierarchical replication as a separate thing. For scenes we only need to match entities once on scene spawning. If we solve this, attaching `ParentSync` to such entity will just work for synchronizing hierarchy changes.

>  It seems like replicon already does some client side buffering

Yes, we buffer mutate messages to apply them only if update message is received.

Before I write detailed information about the implementation, we need to figure out how to properly generate IDs. Looks like the upcoming BSN format won't have `Entity` inside the scene. They rely on scene path. But I need to clarify this, already [asked](https://discord.com/channels/691052431525675048/1264881140007702558/1316444852367392858) in the working group in Discord.

**Comment by blueforesticarus** [comment]
> I would consider hierarchical replication as a separate thing.
I think maybe we are still talking past each other. What I mean is: the problem in my mind is replicating things which, as part of their spawning, spawn child entities which need to be replicated. 

A scene is one example where this is useful. It also has an extra constraint.
1. because scenes are often assets, we want to avoid having needing to write a system specific to the scene. We'd prefer some generic solution using a marker component that can work for all scenes.

Thinking in terms of the current Marker component method for spawning.
1. MyPlayerMarker **hydrates** my player entity by adding **components** (like mesh and material) symmetrically on both client and server. **This works.**
2. MyPlayerMarker2 is the same but also spawns children. This is a problem atm in replicon.

Really at the low level the problem is the replication of entities that (on the client) must be spawned by client code. Replicon currently assumes that anything it replicates, it also spawns (ignoring prespawning). 
If this assumption is removed, we need a way for the replicon client to map entities it didn't spawn. A unique id (hashed to a u64 or something), seems like the obvious choice for the low level. **Any other method of mapping can be implemented as a system which assigns these unique ids.**

**Comment by Shatur** [comment]
> 2. MyPlayerMarker2 is the same but also spawns children. This is a problem atm in replicon.

There is no issue here. You can spawn hierarchies on server with `ParentSync`. And have systems that initializes entities for both client and server. It works exactly like "1".

The scene use case is completely different. You want to have base entities on both client and server and receive updates to these values if they changed. There is no need to introduce custom IDs for hierarchy replication. We need them specially for scenes to know which scene entity belongs to which spawned entity, that's it.

**Comment by blueforesticarus** [comment]
>There is no issue here. You can spawn hierarchies on the server with ParentSync and have systems that initialize entities for both the client and server. It works exactly like "1."

No, there is an asymmetry. The Marker which spawns children (which themselves need replication) must only spawn children on the server.

**Comment by Shatur** [comment]
> ...spawns children that themselves need replication.

Yes, I get this. It's supported. With both spawning on server and pre-spawning on client.

> Which I suppose is potentially more general than hierarchical structures, just any group of entities spawned by the same code.

I can't think of any use case other then scenes. For all other use cases you can use regular replication from server or pre-spawning.

**Comment by blueforesticarus** [comment]
> Yes, I get this. It's supported. With both spawning on server and pre-spawning on client.
I don't get how currently you can accomplish this with the pre-spawning system. It seems like the client *must* spawn the entity first, because it has to give the server the entity ID. 

Maybe I don't actually get how pre-spawning works, it does seem bizarre and roundabout to me.

**Comment by Shatur** [comment]
> It seems like the client must spawn the entity first, because it has to give the server the entity ID.

That's the whole idea of pre-spawning. You spawn entity or entities and notify server about them. Then server spawns them, and client will receive replication as like they were spawned on server. It's useful if you want to hide delay on client, typically needed for prediction for fast-paced games.

If you don't need to spawn on client first - you just spawn things on server. If you also want to replicate the hierarchy - you insert `ParentSync`.

> A) is not symmetric
> B) seems to require spawning on the client first.

Yes, it's completely different from the scene case.
